File: web/src/main/java/top/bettercode/summer/web/AnnotatedUtils.java
package top.bettercode.summer.web;

import java.lang.annotation.Annotation;
import java.util.Set;
import org.springframework.core.annotation.AnnotatedElementUtils;
import org.springframework.web.method.HandlerMethod;

/**
 * @author Peter Wu
 */
public class AnnotatedUtils {


  public static <A extends Annotation> boolean hasAnnotation(HandlerMethod handlerMethod,
      Class<A> annotationType) {
    if (handlerMethod.hasMethodAnnotation(annotationType)) {
      return true;
    } else {
      return AnnotatedElementUtils.hasAnnotation(handlerMethod.getBeanType(), annotationType);
    }
  }

  public static <A extends Annotation> A getAnnotation(HandlerMethod handlerMethod,
      Class<A> annotationType) {
    A annotation = handlerMethod.getMethodAnnotation(annotationType);
    return annotation == null ? AnnotatedElementUtils.getMergedAnnotation(
        handlerMethod.getBeanType(), annotationType) : annotation;
  }

  public static <A extends Annotation> Set<A> getAnnotations(HandlerMethod handlerMethod,
      Class<A> annotationType) {
    Set<A> annotations = AnnotatedElementUtils.getMergedRepeatableAnnotations(
        handlerMethod.getMethod(), annotationType);
    return annotations.isEmpty() ? AnnotatedElementUtils.getMergedRepeatableAnnotations(
        handlerMethod.getBeanType(), annotationType) : annotations;
  }

}

File: tools/tools/src/main/kotlin/top/bettercode/summer/tools/lang/util/AnnotatedUtils.kt
package top.bettercode.summer.tools.lang.util

import org.springframework.core.annotation.AnnotatedElementUtils
import org.springframework.web.method.HandlerMethod

/**
 * @author Peter Wu
 */
object AnnotatedUtils {
    @JvmStatic
    fun <A : Annotation?> hasAnnotation(
            handlerMethod: HandlerMethod,
            annotationType: Class<A>
    ): Boolean {
        return if (handlerMethod.hasMethodAnnotation(annotationType)) {
            true
        } else {
            AnnotatedElementUtils.hasAnnotation(
                    handlerMethod.beanType,
                    annotationType
            )
        }
    }

    @JvmStatic
    fun <A : Annotation?> getAnnotation(
            handlerMethod: HandlerMethod,
            annotationType: Class<A>
    ): A? {
        val annotation = handlerMethod.getMethodAnnotation(annotationType)
        return annotation
                ?: AnnotatedElementUtils.getMergedAnnotation(
                        handlerMethod.beanType, annotationType
                )
    }

    @JvmStatic
    @JvmOverloads
    fun <A : Annotation?> getAnnotations(
            handlerMethod: HandlerMethod,
            annotationType: Class<A>,
            all: Boolean = false
    ): Set<A> {
        val annotations = AnnotatedElementUtils.getMergedRepeatableAnnotations(
                handlerMethod.method, annotationType
        )
        return if (all) {
            val annotations1 = AnnotatedElementUtils.getMergedRepeatableAnnotations(
                    handlerMethod.beanType, annotationType
            )
            annotations + annotations1
        } else {
            if (annotations.isEmpty()) AnnotatedElementUtils.getMergedRepeatableAnnotations(
                    handlerMethod.beanType, annotationType
            ) else annotations
        }
    }
}